networkx/algorithms/dag.py

```diff
@@ -170,10 +180,16 @@ def topological_sort(G):
 
     indegree_map = {v: d for v, d in G.in_degree() if d > 0}
     # These nodes have zero indegree and ready to be returned.
-    zero_indegree = [v for v, d in G.in_degree() if d == 0]
+    zero_indegree = deque(v for v, d in G.in_degree() if d == 0)
+
+    generation = 0
+    if zero_indegree:
+        last_node_in_generation = zero_indegree[-1]
+    else:
+        raise nx.NetworkXUnfeasible("Graph contains a cycle.")
```

Comment:
Is this true? This seems to be a departure from the original implementation, which only raises `if indegree_map`.

Reply:
I believe before the first yield, if there are no zero-indegrees, this can only happen if there is a cycle in the graph. 

```
Note: the above statement is not the only way a cycle can exist, e.g.

B > A
C > A
A > B

This will have a cycle, but still contain a zero-indegree
```
Despite the above, if an error can be confidently raised concerning the infeasibility of a dag in a specific condition, I think it would make sense to do it asap. Especially since identifying that case doesn't add any extra time/space complexity to the algorithm, and reduces unnecessary processing of the dag.

Reply:
👍  I'll take a look over the weekend.

Reply:
so that's what all those tests are for! 😃

Reply:
Updated. Passing all tests run with `pytest networkx/`. Also added a few tests for the new function/topo-sort updates.